spotbugs/src/main/java/edu/umd/cs/findbugs/util/ClassName.java

```diff
@@ -175,9 +175,47 @@ public static String extractSimpleName(@DottedClassName String className) {
      * @return true if it's a valid class name, false otherwise
      */
     public static boolean isValidClassName(String className) {
-        // FIXME: should use a regex
+        return isValidBinaryClassName(className) ||
+                isValidDottedClassName(className) ||
+                isValidArrayFieldDescriptor(className) ||
+                isValidClassFieldDescriptor(className);
+    }
+
+    private static boolean isValidBinaryClassName(String className) {
+        return className.indexOf('.') == -1 &&
+                className.indexOf('[') == -1 &&
+                className.indexOf(';') == -1;
+    }
+
+    private static boolean isValidDottedClassName(String className) {
+        return className.indexOf('/') == -1 &&
+                className.indexOf('[') == -1 &&
+                className.indexOf(';') == -1;
+    }
+
+    private static boolean isValidArrayFieldDescriptor(String className) {
+        return className.startsWith("[") &&
+                (isValidArrayFieldDescriptor(className.substring(1))) ||
+                isValidClassFieldDescriptor(className.substring(1)) ||
+                isValidBaseTypeFieldDescriptor(className.substring(1));
```

Comment:
It's just my personal and unreasonable motivation. JIT could be enough.

```suggestion
        String tail = className.substring(1);
        return className.startsWith("[") &&
                (isValidArrayFieldDescriptor(tail)) ||
                isValidClassFieldDescriptor(tail) ||
                isValidBaseTypeFieldDescriptor(tail);
```

Reply:
I actually thinks this improves readability too, so this might be a good change. :+1:

However, I am concerned about the possibility of throwing an `IndexOutOfBoundsException` if this is called with an empty string (instead of returning false). I have added an `isEmpty()` check in the caller and documentation to mitigate this concern.